nexus/db-queries/src/db/raw_query_builder.rs

```diff
@@ -0,0 +1,195 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Utilities for building string-based queries.
+//!
+//! These largely side-step Diesel's type system,
+//! and are recommended for more complex CTE
+
+use crate::db::pool::DbConnection;
+use diesel::pg::Pg;
+use diesel::query_builder::{AstPass, Query, QueryFragment, QueryId};
+use diesel::sql_types;
+use diesel::RunQueryDsl;
+use std::cell::Cell;
+use std::marker::PhantomData;
+
+// Keeps a counter to "how many bind parameters have been used" to
+// aid in the construction of the query string.
+struct BindParamCounter(Cell<i32>);
+impl BindParamCounter {
+    fn new() -> Self {
+        Self(0.into())
+    }
+    fn next(&self) -> i32 {
+        self.0.set(self.0.get() + 1);
+        self.0.get()
+    }
+}
+
+/// A "trusted" string, which can be used to construct SQL queries even
+/// though it isn't static. We use "trust" to refer to "protection from
+/// SQL injections".
+///
+/// This is basically a workaround for cases where we haven't yet been
+/// able to construct a query at compile-time.
+pub struct TrustedStr(TrustedStrVariants);
+
+impl TrustedStr {
+    /// Explicitly constructs a string, with a name that hopefully
+    /// gives callers some pause when calling this API.
+    ///
+    /// If arbitrary user input is provided here, this string COULD
+    /// cause SQL injection attacks, so each call-site should have a
+    /// justification for "why it's safe".
+    pub fn i_take_responsibility_for_validating_this_string(s: String) -> Self {
```

Comment:
this is wonderful :)

Reply:
I hope it's the right direction, and I'd be stoked to move things more and more into "const fn"s as we can.

I think the choices made by Diesel (e.g. https://docs.diesel.rs/master/diesel/query_builder/struct.SqlQuery.html#method.sql, which just takes anything that impls `AsRef<str>` ) makes this a little too easy to foot-gun via SQL injection when building queries that side-step the norms of Diesel's type system.

I know it's possible to subvert this with `&'static str` too, and I wish there was a way to more explicitly say "compile-time-constructed string", but that's as close a proxy as I think I can get right now.